src/anomalib/models/ai_vad/density.py

```diff
@@ -0,0 +1,295 @@
+        Returns:
+            Tensor: Mean (normalized) distances of input feature vectors to k nearest neighbors in feature bank.
+        """
+        n_neighbors = n_neighbors or self.n_neighbors
+
+        if group:
+            mem_bank = self.memory_bank.copy()
+            mem_bank.pop(group)
+        else:
+            mem_bank = self.memory_bank
+
+        mem_bank_tensor = torch.vstack(list(mem_bank.values()))
+
+        distances = self._nearest_neighbors(mem_bank_tensor, features, n_neighbors=n_neighbors)
+
+        if normalize:
+            distances = self._normalize(distances)
+
+        return distances.mean(axis=1)
+
+    @staticmethod
+    def _nearest_neighbors(feature_bank: Tensor, features: Tensor, n_neighbors: int = 1):
+        """Perform the KNN search.
+
+        Args:
+            feature_bank (Tensor): Feature bank used for KNN search.
+            features (Ternsor): Input features.
+            n_neighbors (int): Number of neighbors used in KNN search.
+        Returns:
+            Tensor: Distances between the input features and their K nearest neighbors in the feature bank.
+        """
+        distances = torch.cdist(features, feature_bank, p=2.0)  # euclidean norm
+        if n_neighbors == 1:
+            # when n_neighbors is 1, speed up computation by using min instead of topk
+            distances, _ = distances.min(1)
+            return distances.unsqueeze(1)
+        else:
+            distances, _ = distances.topk(k=n_neighbors, largest=False, dim=1)
+        return distances
+
+    def _compute_normalization_statistics(self):
+        """Compute min-max normalization statistics while taking the group into account."""
+        for group, features in self.memory_bank.items():
+            nns = self.predict(features, group, normalize=False)
+            self.normalization_statistics.update(nns)
+
+        self.normalization_statistics.compute()
+
+    def _normalize(self, distances: Tensor):
+        """Normalize distance predictions.
+
+        Args:
+            distances (Tensor): Distance tensor produced by KNN search.
+        Returns:
+            Tensor: Normalized distances.
+        """
+        return (distances - self.normalization_statistics.min) / (
+            self.normalization_statistics.max - self.normalization_statistics.min
+        )
+
+
+class GMMEstimator(BaseDensityEstimator):
+    """Density estimation based on Gaussian Mixture Model.
+
+    Args:
+        n_components (int): Number of components used in the GMM.
+    """
+
+    def __init__(self, n_components: int = 2) -> None:
+        super().__init__()
+
+        self.gmm = GaussianMixture(n_components=n_components, random_state=0)
+        self.memory_bank: list[Tensor] | Tensor = []
+
+        self.normalization_statistics = MinMax()
+
+    def update(self, features: Tensor, _group: Any = None):
+        """Update the feature bank."""
+        self.memory_bank.append(features)
```

Comment:
below is the new recommended way to ignore unused variables

```suggestion
    def update(self, features: Tensor, group: Any = None):
        """Update the feature bank."""
        del group    # This variable is not needed.
        self.memory_bank.append(features)
```